scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -139,6 +140,9 @@ object ExecutionOptimizationRules {
       def apply(in: (A, (B, C))) =
         (in._2._1, (in._1, in._2._2))
     }
+    case class ComposedFn[A1, A2, A, B1, B2, B](fn1: ((A1, A2)) => A, fn2: ((B1, B2)) => B) extends Function1[((A1, B1), (A2, B2)), (A, B)] {
```

Comment:
This seems a bit complex. Do we need the tuples? Isn’t this just A => B with C => D to make ((A, C)) => (B, D) feels like the extra tuple types are distracting here.

Reply:
Not exactly because it's also do twist around the arguments.